### PR TITLE
preventing Nan error with adding epsilon(0.00000001) trick

### PR DESCRIPTION
**Pull request by imcomking**

### CODE REVIEW

File: 08_4_softmax_pg_pong.py [hunk 1]
```diff
@@ -53,7 +53,7 @@
 
 # Loss function
 # Sum (Ai*logp(yi|xi))
-log_lik = -Y * tf.log(action_pred)
+log_lik = -Y * (tf.log(action_pred)+0.0000000001)
```
Comment (kkweon) [inline 08_4_softmax_pg_pong.py]: nan + any number is still nan though.

Shouldn't you add inside the log?
`tf.log(action_pred + 1e-8)`

**Comment by hunkim** [comment]
Is it working? Do you have loss/reward graphs?

File: 08_4_softmax_pg_pong.py (hunk 1 above)
Comment (imcomking) [inline 08_4_softmax_pg_pong.py]: Sorry. I made a mistake like you said. It should be like this.

log_lik = -Y * (tf.log(tf.clip_by_value(action_pred, 1e-10 , 1.0)))

**Comment by hunkim** [comment]
Cool. Now we beat the pong? Thanks.

**Comment by imcomking** [comment]
Yes I have the graphs.

#1. Sum of rewards per episode graph which is smooted with 0.74 using tensorboard. I think this is Slowly increasing.
image

# Loss graph. Loss converges to zero.
image


#3. You can see the convergence of "action_pred vector". They know what to do or no
image

**Comment by hunkim** [comment]
How long did it take? I'm not sure the reward is increasing though. We will
see. Thanks.

**Comment by imcomking** [comment]
I trained the agent with 1 titan X, for almost 1 weeks.

And I think the training is really slow because of some implementation issues.

### 1. The learning(optimization) is executed only few times.
In the present implementation, the 'train' operation is called whenever the episode is finished.
Like follows.

image

This means the S.G.D. occurs only 1000 time for 1 weeks of training. This is really few updates of agent's strategy for beating the opponent.

### 2. The p.g. is really inefficient for the data.
The current codes throw away the past experiences because the p.g. is on-policy method.
We have fix this.

### 3. The task is too hard to success.
If we do not fix the first and second issue, this pong is impossible to clear.

**Comment by kkweon** [comment]
There are some flaws in this code.
```text
log_lik = (n_samples, n_actions)
advantages = (n_samples, 1)
```
So, when we compute this
```python
loss = tf.reduce_mean(tf.reduce_sum(log_lik * advantages, axis=1))
```

```text
log_lik = (n_samples, n_actions)
advantages = (n_samples, n_actions) # <- (n_samples, 1)
```
and reduce sum adding all the probability with wrong advantage.

**Comment by imcomking** [comment]
Actually, your pointing out doesn't matter.

The "log_lik" is always multiplied with "Y" like this.
`log_lik = -Y * (tf.log(tf.clip_by_value(action_pred, 1e-10 , 1.0)))`

And the Y is one-hot coded action vector. like this.
image
image

image

So, if we do this ,'tf.reduce_sum(log_lik * advantages, axis=1)' , the result will be same with your suggestions.
'log_lik.shape = (n_samples, n_actions)
advantages.shape = (n_samples, n_actions)'

'loss = tf.reduce_mean(tf.reduce_sum(log_lik * advantages, axis=1))'

If I am wrong, please correct me.

**Comment by kkweon** [comment]
@imcomking yeah you're right. my bad. 
Didn't normalizing rewards/advantages help?

**Comment by imcomking** [comment]
"I'm not sure of adding 4 images too because the action is already repeated several times by the openai gym."

No, the action repeat is not implemented is this code.
If we want to the action repeat, a.k.a n-frame skipping, we have to implement like this.
image

**Comment by imcomking** [comment]
@kkweon 
"Didn't normalizing rewards/advantages help?"

I didn't test the effect of normalizing, I will try to that thank you.

**Comment by kkweon** [comment]
This is from the [gym document](https://gym.openai.com/envs/Pong-v0)
> action is repeatedly performed for a duration of k frames, where k is uniformly sampled from {2,3,4}

**Comment by imcomking** [comment]
Oh sorry, you're right, there is also default n-frame skipping.

But the state(image) is returned after the "default n-frame skipping of gym".

So using the 4 images is not wrong.
